### PR TITLE
MISP Feed Limit Fix

### PR DESCRIPTION
**Pull request by AradCarmi**
## Related Issues
relates: https://jira-dc.paloaltonetworks.com/browse/XSUP-40038

## Description
Fixed an issue with the limit parameter and re-fetched the same indicators.

### CODE REVIEW

File: Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py [hunk 1]
```diff
@@ -529,31 +529,60 @@ def fetch_attributes_command(client: Client, params: Dict[str, str]):
...
         for iter_ in batch(indicators, batch_size=2000):
             demisto.createIndicators(iter_)
         params_dict['page'] += 1
-        last_run = search_query_per_page['response']['Attribute'][-1]['timestamp']
+
+        # If the indicators created successfully and the new indicator timestamp is bigger than last run timestamp,
+        # update the last run timestamp and last indicator value
+        if ((last_run_timestamp and latest_indicator_timestamp
+             and latest_indicator_timestamp > last_run_timestamp)
+                or not last_run_timestamp):
```
Comment (dantavori) [inline Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py:569]:
```suggestion
        if not last_run_timestamp or latest_indicator_timestamp > last_run_timestamp:
```

File: Packs/FeedMISP/ReleaseNotes/1_0_38.md [hunk 2]
```diff
@@ -0,0 +1,7 @@
+
+#### Integrations
+
+##### MISP Feed
+
+- Fixed an issue where the **Max. indicators per fetch** parameter failing to halt the ***fetch-indicators*** process.
```
Comment (dantavori) [inline Packs/FeedMISP/ReleaseNotes/1_0_38.md:6]: need to rephrase this RN, why is a `Max. indicators per fetch` parameter responsible for halting fetch-indicators? I understand what you mean but think from a customer's perspective

**Review by dantavori — APPROVED** [review]
I'm approving but see notes
@MLainer1 I want you to review this as well - will the changes here be safe for all customers? focus on:
1. how we use the page from the last run, and sort each page separately
2. taking into account different MISP instances

File: Packs/FeedMISP/Integrations/FeedMISP/FeedMISP_test.py [hunk 3]
```diff
@@ -375,3 +377,92 @@ def test_parsing_user_query_timestamp_deprecated():
...
+        'filters': {'category': ['Payload delivery']},
+    }
+    mocked_last_run = {"timestamp": "1607517728", "last_indicator_value": "test1"}
+    mocker.patch.object(demisto, 'getLastRun', return_value=mocked_last_run)
+    setLastRun_mocked = mocker.patch.object(demisto, 'setLastRun')
+    mocker.patch.object(demisto, 'createIndicators')
+    fetch_attributes_command(client, params_dict)
+    indicators = demisto.createIndicators.call_args[0][0]
+    # The last ignored indicator will be re-fetched but the new last run will proceed.
+    assert len(indicators) == 2
```
[the remark this answers is not in the document]
Comment (AradCarmi) [reply Packs/FeedMISP/Integrations/FeedMISP/FeedMISP_test.py:467]: The UT simulates a case where we get the same indicators over and over when suddenly, a new indicator is being fetched from the API. That way we will get test1 again (as our last time is still the last indicator time) but this cycle we want to see that the last run will be updated to the new last run.
I will change the remark to be more clear.

**Review by MLainer1 — COMMENTED** [review]
🆒

File: Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py [hunk 4]
```diff
@@ -514,6 +512,51 @@ def get_attributes_command(client: Client, args: Dict[str, str], params: Dict[st
     )
 
 
+def update_candidate(last_run: dict, last_run_timestamp: str, last_run_value: str,
+                     latest_indicator_timestamp: str, latest_indicator_value: str):
```
Comment (dantavori) [inline Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py:516]:
```suggestion
def update_candidate(last_run: dict, last_run_timestamp: int, last_run_value: str,
                     latest_indicator_timestamp: int, latest_indicator_value: str):
```

File: Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py [hunk 5]
```diff
@@ -514,6 +512,51 @@ def get_attributes_command(client: Client, args: Dict[str, str], params: Dict[st
...
+        last_run_timestamp: the timestamp of the last run.
+        last_run_value: the indicator value of the last run.
+        latest_indicator_timestamp: the timestamp of the latest indicator.
+        latest_indicator_value: the value of the latest indicator.
+    """
+    candidate_timestamp = last_run.get('candidate_timestamp') or last_run_timestamp
+    candidate_indicator_value = last_run.get('candidate_value') or last_run_value
+    if latest_indicator_timestamp > candidate_timestamp:
+        candidate_timestamp = latest_indicator_timestamp
+        candidate_indicator_value = latest_indicator_value
```
Comment (dantavori) [inline Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py:531]:
```suggestion
    if latest_indicator_timestamp > candidate_timestamp:
        last_run['candidate_timestamp'] = latest_indicator_timestamp
        last_run['candidate_value'] = latest_indicator_value
```

File: Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py (hunk 4 above)
Comment (AradCarmi) [reply Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py:516]: why int? the value is retrieved from the last run as a str.

File: Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py [hunk 6]
```diff
@@ -514,6 +512,51 @@ def get_attributes_command(client: Client, args: Dict[str, str], params: Dict[st
...
+    last_run['last_indicator_timestamp'] = last_run_timestamp
+    last_run['last_indicator_value'] = last_run_value
+    last_run['candidate_timestamp'] = candidate_timestamp
+    last_run['candidate_value'] = candidate_indicator_value
```
[the remark this answers is not in the document]
Comment (dantavori) [reply Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py:535]: just a suggestion to shorten the code, I will do a clearer unified suggestion for everything in this method below

Comment (dantavori) [inline Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py:535]:
```suggestion
    last_run['candidate_timestamp'] = last_run.get('candidate_timestamp') or last_run_timestamp
    last_run['candidate_value'] = last_run.get('candidate_value') or last_run_value
    if latest_indicator_timestamp > last_run['candidate_timestamp']:
        last_run['candidate_timestamp'] = latest_indicator_timestamp
        last_run['candidate_value'] = latest_indicator_value
```

Comment (AradCarmi) [reply Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py:535]: I'm not sure inserting the value to the 'candidate_timestamp' in the last value will clarify the code. I prefer to leave it as is.

**Review by dantavori — APPROVED** [review]

**Pull request merged**